### PR TITLE
Workaround for a GHC bug with -XTemplateHaskell and ghci (fixes #303)

### LINKED ISSUE #303: doctest's spec tests failing with GHC 9
There are some test failures when building with GHC 9. Eg:

```
# stack8.10.yaml
resolver: lts-18.0
packages:
- .
```

```
# stack.yaml
resolver: nightly-2021-07-16
packages:
- .
```

With ghc 8:
```
$ stack test doctest:test:spec --stack-yaml stack8.10.yaml
...
Finished in 82.0428 seconds
170 examples, 0 failures

doctest> Test suite spec passed
```
With ghc 9:
```
$ stack test doctest:test:spec
...
Failures:

  test/MainSpec.hs:94:7: 
  1) Main.doctest template-haskell
       template-haskell ["Foo.hs"]
       expected: Examples: 2  Tried: 2  Errors: 0  Failures: 0
        but got: Examples: 2  Tried: 2  Errors: 0  Failures: 1

  To rerun use: --match "/Main/doctest/template-haskell/"

  test/MainSpec.hs:148:7: 
  2) Main, doctest as a runner for QuickCheck properties, runs $setup before each property
       property-setup ["Foo.hs"]
       expected: Examples: 3  Tried: 3  Errors: 0  Failures: 0
        but got: Examples: 3  Tried: 3  Errors: 0  Failures: 1

  To rerun use: --match "/Main/doctest as a runner for QuickCheck properties/runs $setup before each property/"

  test/MainSpec.hs:177:7: 
  3) Main, doctest (regression tests), template-haskell-bugfix
       template-haskell-bugfix ["Main.hs"]
       expected: Examples: 2  Tried: 2  Errors: 0  Failures: 0
        but got: Examples: 2  Tried: 2  Errors: 0  Failures: 1

  To rerun use: --match "/Main/doctest (regression tests)/template-haskell-bugfix/"

  test/RunSpec.hs:108:5: 
  4) Run.doctest can deal with potentially problematic GHC options
       uncaught exception: ExitCode
       ExitFailure 1

  To rerun use: --match "/Run/doctest/can deal with potentially problematic GHC options/"

  test/RunSpec.hs:128:11: 
  5) Run.doctestWithOptions, on parse error, prints a useful error message
       expected: "\nFoo.hs:6:1: error:\n    parse error (possibly incorrect indentation or mismatched brackets)\n"
        but got: "\n\ESC[;1mFoo.hs:6:1: \ESC[;1m\ESC[31merror:\ESC[0m\ESC[0m\ESC[;1m\ESC[0m\ESC[0m\ESC[;1m\n    parse error (possibly incorrect indentation or mismatched brackets)\ESC[0m\ESC[0m\n"

  To rerun use: --match "/Run/doctestWithOptions/on parse error/prints a useful error message/"

Randomized with seed 2096325755

Finished in 90.4501 seconds
170 examples, 5 failures
```

### PR DESCRIPTION
**Pull request by sol**
Related GHC bug: https://gitlab.haskell.org/ghc/ghc/-/issues/20670

### CODE REVIEW

File: doctest.cabal [hunk 1]
```diff
@@ -1,10 +1,8 @@
-cabal-version: 1.12
+cabal-version: 2.0
 
 -- This file has been generated from package.yaml by hpack version 0.34.4.
 --
 -- see: https://github.com/sol/hpack
---
--- hash: d72341386fa5019d836d102ce53149841ad11aa7bd786c3ba9b8cc28ca3c9046
```
Comment (andreasabel) [inline doctest.cabal:6]: Does this mean that `doctest.cabal` is no longer created by `hpack`?

File: src/Interpreter.hs [hunk 2]
```diff
@@ -55,13 +56,17 @@ withInterpreter flags action = do
   let
     args = flags ++ [
         "--interactive"
+      , xTemplateHaskell
 #if __GLASGOW_HASKELL__ >= 802
```
Comment (andreasabel) [inline src/Interpreter.hs:60]: What is the meaning of such an `#if`?  The `#if` is evaluated when `doctest` is compiled, and we use such `#if`s all over the place to ensure compilation with different GHCs.  But this `#if` does not make to seem a difference on the _compilation_ of `doctest`, but when I _run_ `doctest`.
Could it be a problem when I run `doctest` with a different GHC version than it was compiled with?

**Review by andreasabel — COMMENTED** [review]
This is already merged, but I anyway have 2 questions (see above).

File: doctest.cabal (hunk 1 above)
Comment (sol) [reply doctest.cabal:6]: No, it's still created by `hpack`.  It merely means that `hpack` will overwrite the file, even if it was modified manually.

File: src/Interpreter.hs (hunk 2 above)
Comment (sol) [reply src/Interpreter.hs:60]: `doctest` uses both `ghc` (the library) and `ghc --interactive` (the executable).  The former is used to extract Haddock comments, the later is used to check examples.  If we would only depend on the executable, we could just use whatever `ghc` is on the `PATH`.  But with `ghc` the library we can only depend on one version (specifically, the version we compile `doctest` with).  For that reason we fix both, `ghc` the library and `ghc` the executable to the `ghc` version we compile against at compile time.  The library version is fixed simply by depending on it. To fix the executable version we use `ghc-paths`.

Comment (andreasabel) [reply src/Interpreter.hs:60]: Ah, this makes sense if `doctest` sits within a per-GHC buildtree.  
(For some reason, I have `doctest` in my PATH, but that does not make sense in such a scenario.)

**Pull request merged**